server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,186 @@
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+// TODO(ES-5727): add a retry mechanism to TransportGetFromTranslogAction
+public class TransportGetFromTranslogAction extends HandledTransportAction<
+    TransportGetFromTranslogAction.Request,
+    TransportGetFromTranslogAction.Response> {
+
+    public static final String NAME = "internal:data/read/get_from_translog";
+    public static final Logger logger = LogManager.getLogger(TransportGetFromTranslogAction.class);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, Request::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final GetRequest getRequest = request.getRequest();
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        assert indexShard.routingEntry().isPromotableToPrimary();
+        assert getRequest.realtime();
+        Engine engine = indexShard.getEngineOrNull();
+        if (engine == null) {
+            listener.onFailure(new AlreadyClosedException("engine closed"));
+            return;
+        }
+        var get = new Engine.Get(getRequest.realtime(), getRequest.realtime(), getRequest.id()).version(getRequest.version())
+            .versionType(getRequest.versionType());
+        assert Objects.equals(get.uid().field(), IdFieldMapper.NAME) : get.uid().field();
+        ActionListener.completeWith(listener, () -> {
+            // TODO: rather than relying on a lookup first and then doing the Get, pass a flag to GetShardService#get
+            //  that will only attempt retrieving the doc from the Translog and otherwise returns NOT_EXISTS.
+            if (engine.isInVersionMap(get.uid().bytes())) {
+                var result = indexShard.getService()
+                    .get(
+                        getRequest.id(),
+                        getRequest.storedFields(),
+                        getRequest.realtime(),
+                        getRequest.version(),
+                        getRequest.versionType(),
+                        getRequest.fetchSourceContext(),
+                        getRequest.isForceSyntheticSource()
+                    );
+                return new Response(result, -1);
+            }
+            long segmentGeneration = engine.commitStats().getGeneration();
+            return new Response(null, segmentGeneration);
+        });
+    }
+
+    public static class Request extends ActionRequest {
```

Comment:
Let us also add a serialization test for this.

Reply:
Thanks for pointing out that `GetRequest` prevents this. For now, let us omit this.